primitives/runtime/src/offchain/storage_lock.rs

```diff
@@ -0,0 +1,184 @@
+//!         acc.set(v);
+//!    } else {
+//!         // the lock duration expired
+//!    }
+//! }
+//! ```
+
+use crate::offchain::storage::StorageValueRef;
+use sp_core::offchain::{Duration, Timestamp};
+use sp_io::offchain;
+
+/// Default expirey duration in milliseconds.
+const STORAGE_LOCK_DEFAULT_EXPIREY_DURATION: u64 = 30_000;
+
+/// Snooze duration before attempting to lock again in ms.
+const STORAGE_LOCK_PER_CHECK_ITERATION_SNOOZE: u64 = 100;
+
+/// A persisted guard state.
+///
+/// An in DB persistent mutex for multi access items which are modified
+/// i.e. vecs or sets.
+pub struct StorageLock<'a> {
+	// A storage value ref which defines the DB entry representing the lock.
+	value_ref: StorageValueRef<'a>,
+	// `None` implies it was already released by `fn unlock(..)`
+	locked_until: Option<Timestamp>,
+}
+
+impl<'a> StorageLock<'a> {
+	/// Create a new storage lock with [default expirey duration](Self::STORAGE_LOCK_DEFAULT_EXPIREY_DURATION).
+	pub fn new<'k>(key: &'k [u8]) -> Self
+	where
+		'k: 'a,
+	{
+		Self {
+			value_ref: StorageValueRef::<'a>::persistent(key),
+			locked_until: None,
+		}
+	}
+
+	#[inline]
+	fn try_lock_inner(&mut self, duration: Duration) -> Result<(), Option<Timestamp>> {
+		let now = offchain::timestamp();
+		let expires_at = now.add(duration);
+		let res = self.value_ref.mutate(
+			|s: Option<Option<Timestamp>>| -> Result<Timestamp, Option<Timestamp>> {
+				match s {
+					// no lock set, we can safely acquire it
+					None => Ok(expires_at),
+					// lock is set, but it's old. We can re-acquire it.
+					Some(Some(current_good_until)) if current_good_until < now => Ok(expires_at),
+					// lock is present and is still active
+					Some(Some(current_good_until)) => Err(Some(current_good_until)),
+					_ => Err(None),
+				}
+			},
+		);
+		match res {
+			Ok(Ok(_)) => {
+				self.locked_until = Some(expires_at);
+				Ok(())
+			}
+			Ok(Err(timestamp)) => Err(Some(timestamp)), // failed to set the new value, but could read the current
+			Err(e) => Err(e),                           // forward the remaining value
+		}
+	}
+
+	/// Attempt to lock the storage entry.
+	///
+	/// Returns a lock guard on success, otherwise an error containing `None` in
+	/// case the mutex was already unlocked before, or if the lock is still held
+	/// by another process `Err(Some(expiration_timestamp))`.
+	pub fn try_lock<'b>(&'b mut self) -> Result<StorageLockGuard<'a, 'b>, Option<Timestamp>>
+	where
+		'a: 'b,
+	{
+		if self.locked_until.is_none() {
+			match self.try_lock_inner(Duration::from_millis(STORAGE_LOCK_DEFAULT_EXPIREY_DURATION))
+			{
+				Ok(_) => Ok(StorageLockGuard::<'a, 'b> { lock: Some(self) }),
+				Err(e) => Err(e),
+			}
+		} else {
+			Err(self.locked_until)
+		}
+	}
+
+	/// Try grabbing the lock until its expiry is reached.
+	///
+	/// Returns an error if the lock expired before it could be cought
+	pub fn spin_lock<'b, 'c>(&'b mut self) -> Result<StorageLockGuard<'a, 'c>, ()>
```

Comment:
IMHO `spin_lock` should not really fail after the default expiry time. I feel we mix two things here:
1. The LONGEST time the lock is acquired for.
 2. The time that the caller is willing to wait for the lock to be acquired.

TBH I imagine 3 different cases here (although the last one doesn't make sense with block-based locks that much, only if we decide to have time-based locks):
1. `try_lock` - either lock instantly or not (same as `Mutex::try_lock`)
2. `spin_lock` - try to acquire a lock blocking forever. `spin_` just emphasizes that in OCW context (single thread) it's going to spin. (same as `Mutex::lock`)
3. `spin_lock_with_deadline` - try to acquire a lock, but fail if deadline is reached. That's an API that matches all other HTTP apis that we have (deadline-based).

I feel that API proposed in this PR is not really justified with use cases from the codebase and is a bit arbitrary.

Reply:
The PR is a code extraction + some cleanup of #5773 / what we talked about earlier.

I can adjust the API as desired and make sure to extract two constants instead of one.

Reply:
Removed the explicit timeout when spinning. Eventually every spin lock will now succeed grabbing the lock.